Bugfix/dnstap out of sync

wtoorop opened the pull request:
Recovery of dnstap collector getting out of sync with the serve childs.
TODO:

- [x] respawn dnstap collector on reload without closing sockets like the serve childs (otherwise we leak memory on zone updates because the dnstap collector has old version of database...)

wtoorop committed:
Exit dnstap_collector on EFAULT or EINVAL

wtoorop committed:
Respwan dnstap collector on dnstap exit and..

exit dnstap collector when detecting out of sync

wtoorop committed:
Initialize collector without zones

wtoorop committed:
Fix memory leak with respawning dnstap collector

Once a dnstap collector has crashed or gotten out of sync at some point, it
exits and is respawn, but has the zones content in memory then.  This commit
makes sure that after that happened, a new dnstap collector is forked with each
reload so that the memory will not diverge from the serve childs (which will
cause a gradual memory leak).

TODO: Investigate whether it would be worthwhile to reuse the dt_env and/or the
fd_send and fd_recv when respawning a dnstap collector on reload.

wcawijngaards reviewed (approved):
Looks good, respawn every reload keeps track of the zones in memory. Could also free that data in the new dnstap-collector process, but this may work just as well and could be simpler.

wtoorop committed:
Update Ubuntu cirrus tests to gcc-11 and clang-12

wtoorop committed:
Extra defense against double dnstap logging from Wouter

The service remaining tcp routine is already protected by the set of no more tcp queries on the event handler functions so it should never call dnstap log routines, but this is even more explicit to stop double writes to the logs. If we still want to log from the service remaining tcp, need to set up some pipe for that.

wtoorop committed:
Swap pipe fds to dnstap collector on reload

so old serve childs and new serve childs never write to the same pipes

wcawijngaards reviewed (approved):
Looks good to me.

wtoorop commented:
Closing in favor of #182
